ARROW-17932: [C++] Implement streaming RecordBatchReader for JSON (#14355)
apache/arrow: docs/source/cpp/json.rst

Reading JSON files
==================

Line-separated JSON files can either be read as a single Arrow Table
with a :class:`~TableReader` or streamed as RecordBatches with a
:class:`~StreamingReader`.

Both of these readers require an :class:`arrow::io::InputStream` instance
representing the input file. Their behavior can be customized using a
combination of :class:`~ReadOptions`, :class:`~ParseOptions`, and
other parameters.

.. seealso::
   :ref:`JSON reader API reference <cpp-api-json>`.

TableReader
===========

> Reads an entire file in one shot as a :class:`~arrow::Table`. Each

Reviewer: ```suggestion
:class:`~TableReader` reads an entire file in one shot as a :class:`~arrow::Table`. Each
```

      }
   }

StreamingReader
===============

> Reads a file incrementally in fixed-size blocks, each yielding a

Reviewer: ```suggestion
:class:`~StreamingReader` reads a file incrementally from blocks of a roughly equal byte size, each yielding a
```

:class:`~arrow::RecordBatch`. Each independent JSON object in a block
is converted to a row in the output batch.

All batches adhere to a consistent :class:`~arrow:Schema`, which is
derived from the first loaded batch. Alternatively, an explicit schema
may be passed via :class:`~ParseOptions`.

.. code-block:: cpp

   #include "arrow/json/api.h"

   {
      // ...
      auto read_options = arrow::json::ReadOptions::Defaults();
      auto parse_options = arrow::json::ParseOptions::Defaults();

      std::shared_ptr<arrow::io::InputStream> stream;
      auto result = arrow::json::StreamingReader::Make(stream,
                                                       read_options,
                                                       parse_options);
      if (!result.ok()) {
         // Handle instantiation error
      }
      std::shared_ptr<arrow::json::StreamingReader> reader = *result;

>       std::shared_ptr<arrow::RecordBatch> batch;

Reviewer: I would move this declaration inside the loop, which will avoid holding to memory too long:
```c++
         std::shared_ptr<arrow::RecordBatch> batch = *maybe_batch;
         // Operate on each batch...
```

      for (arrow::Result<std::shared_ptr<arrow::RecordBatch>> maybe_batch : *reader) {

>          if (!result.ok()) {

Reviewer: ```suggestion
         if (!maybe_batch.ok()) {
```
